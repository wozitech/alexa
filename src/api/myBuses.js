'use strict';

import { getTflApiSecret } from '../aws/secrets';
import { nextBusTo } from '../model/tfl.api';

export const handler = async (event, context, callback) => {
  var arnList = (context.invokedFunctionArn).split(":");
  var lambdaRegion = arnList[3];
  
  try {
      if ('undefined' === typeof process.env.TFL_API_SECRET_ID) {
        throw new Error('Missing env variable for TFL_API_SECRET_ID');
      }

      var tflApiDetails = await getTflApiSecret(lambdaRegion,
                                                process.env.TFL_API_SECRET_ID);

      if (typeof event.destination === 'undefined') event.destination = 'Clapham';

      const nextBuses = await nextBusTo(event.destination, tflApiDetails);

      const response = {
        statusCode: 200,
        body: JSON.stringify({
          message: `WOZiTech nextBusto ${event.destination}`,
          event
        }),
      };

      if (nextBuses.endpoint === null) {
        // the destination is unknown
        response.code = 501;
        response.err = 'Destination unknown';
      }
      if (nextBuses.arrivals) {
        response.arrivals = nextBuses.arrivals;
      }

      return response;
  } catch (err) {
      callback(new Error(err));
  }
};
