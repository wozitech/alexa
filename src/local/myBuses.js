const myHello = require('../../lib/myBuses');

const WHEN_IS_MODE=1;
const HOW_LONG_MODE=2;

// stripped out to just the request
const whenIsAlexaExampleEvent = {
    "version": "1.0",
    "request": {
        "type": "IntentRequest",
        "requestId": "amzn1.echo-api.request.833e1635-eda6-4fb5-ae8b-f06e0453c135",
        "timestamp": "2018-11-27T14:48:29Z",
        "locale": "en-GB",
        "intent": {
            "name": "whenIs",
            "confirmationStatus": "NONE",
            "slots": {
                "Destination": {
                    "name": "Destination",
                    "value": "Clapham",
                    "resolutions": {
                        "resolutionsPerAuthority": [
                            {
                                "authority": "amzn1.er-authority.echo-sdk.amzn1.ask.skill.2ba78764-0a67-481f-907a-3f7c08287aeb.LIST_OF_DESTINATIONS",
                                "status": {
                                    "code": "ER_SUCCESS_MATCH"
                                },
                                "values": [
                                    {
                                        "value": {
                                            "name": "Clapham",
                                            "id": "5a0251c093f8679f79d2fc9d474dd768"
                                        }
                                    }
                                ]
                            }
                        ]
                    },
                    "confirmationStatus": "NONE",
                    "source": "USER"
                }
            }
        },
        "dialogState": "STARTED"
    }
};

// stripped out to just the request
const howLongAlexaExampleEvent = {
    "version": "1.0",
    "request": {
        "type": "IntentRequest",
        "requestId": "amzn1.echo-api.request.3d2fe165-8a8f-425d-b286-5dc4a0fa5e4d",
        "timestamp": "2018-11-27T15:38:54Z",
        "locale": "en-GB",
        "intent": {
            "name": "howLong",
            "confirmationStatus": "NONE",
            "slots": {
                "Destination": {
                    "name": "Destination",
                    "value": "Crystal Palace",
                    "resolutions": {
                        "resolutionsPerAuthority": [
                            {
                                "authority": "amzn1.er-authority.echo-sdk.amzn1.ask.skill.2ba78764-0a67-481f-907a-3f7c08287aeb.LIST_OF_DESTINATIONS",
                                "status": {
                                    "code": "ER_SUCCESS_MATCH"
                                },
                                "values": [
                                    {
                                        "value": {
                                            "name": "Crystal Palace",
                                            "id": "5a0251c093f8679f79d2fc9d474dd768"
                                        }
                                    }
                                ]
                            }
                        ]
                    },
                    "confirmationStatus": "NONE",
                    "source": "USER"
                }
            }
        },
        "dialogState": "STARTED"
    }
};

const myBuses = async (requestMode) => {
    try {
        console.log("About to call myBuses handler");
        var handlerEvent=null;
        switch (requestMode) {
            case WHEN_IS_MODE:
                handlerEvent = whenIsAlexaExampleEvent;
                break;

            case HOW_LONG_MODE:
                handlerEvent = howLongAlexaExampleEvent;
                break;
        }

        const returnVal = await myHello.handler(
            handlerEvent,
            { invokedFunctionArn : 'arn:aws:lambda:eu-west-2:accountid:function:wozitech-alexa-skills-dev-myBuses' },
            (err, data) => {
                if (err) console.error(err);
                console.log("LOCAL: My results: ", data);
            }
        );
        console.log("MyBuses.Handler returned: ", returnVal);
    } catch (err) {
        console.error("Caught local error: ", err);
    }
}

process.env.TFL_API_SECRET_ID = 'TFL_API_Portal';

myBuses(HOW_LONG_MODE);
myBuses(WHEN_IS_MODE);