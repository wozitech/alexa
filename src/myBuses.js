'use strict';

import AWS  from 'aws-sdk';

export const handler = async (event, context, callback) => {
  var arnList = (context.invokedFunctionArn).split(":");
  var lambdaRegion = arnList[3];
console.log("DEBUG: Inside myBuses.handler")  
  
  var secrets = new AWS.SecretsManager({
      region: lambdaRegion
  });
  
  try {
      if ('undefined' === typeof process.env.TFL_API_SECRET_ID) {
        throw new Error('Missing env variable for TFL_API_SECRET_ID');
      }

console.log("DEBUG: Attempting to fetch secret")  
      var tflApiSecret = await secrets.getSecretValue({SecretId: process.env.TFL_API_SECRET_ID}).promise();

console.log("DEBUG: tflApiSecret is: ", tflApiSecret);

      if (typeof tflApiSecret.SecretString !== 'undefined') {
        var tflApiDetails = JSON.parse(tflApiSecret.SecretString);

        if (typeof tflApiDetails == 'undefined' ||
            typeof tflApiDetails.tfl_api_app_id == 'undefined' ||
            typeof tflApiDetails.tfl_api_app_key == 'undefined') {
          throw new Error('Unexpected TFL API credentials');
        }

        console.log("DEBUG: tflApiDetails are: ", tflApiDetails);
  
        const response = {
          statusCode: 200,
          body: JSON.stringify({
            message: 'Go Serverless v1.0! Your function executed successfully!',
            input: event,
            api: tflApiDetails
          }),
        };
        return response;

      } else {
        throw new Error('Unknown TFL API');
      }

  } catch (err) {
      callback(new Error(err));
  }
};
