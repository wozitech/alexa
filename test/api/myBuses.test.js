import { handler } from '../../src/api/myBuses';
import AWS  from 'aws-sdk';
import * as TFL from '../../src/model/tfl.api';

const getSecretValueMock = jest.fn();
const nextBusToMock = jest.fn();

AWS.SecretsManager = jest.fn( () => ({
    getSecretValue : getSecretValueMock
}));

TFL.nextBusTo = nextBusToMock;
nextBusToMock.mockReturnValue({
    endpoint: 'https://mocked/endpoint',
    arrivals: [
        '2018-11-27T11:13:33Z',
        '2018-11-27T09:44:02Z',
        '2018-11-27T11:19:05Z',
    ]
});

describe('The myBuses handler', () => {
    describe('Calling the handler', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });
    
        const theEvent = {
        };
        const theContext = {
            invokedFunctionArn : 'arn:aws:lambda:eu-west-2:accountid:function:wozitech-alexa-skills-dev-myBuses'
        };
        const mockCallback = jest.fn();

        it('should return exception when env var TFL_API_SECRET_ID is undefined ', async () => {
            try {
                await handler(theEvent, theContext, mockCallback);
            } catch (err) {
                expect(err).toEqual(new Error('Missing env variable for TFL_API_SECRET_ID'));
            }
        });

        it ('should return exception when the fetched secret is not as expected', async () => {
            getSecretValueMock.mockReturnValueOnce({
                promise: () => ({
                    empty: true
                })
            });
            try {
                const tglApiEnv = 'TFL_API_Portal';
                process.env.TFL_API_SECRET_ID = tglApiEnv;
                await handler(theEvent, theContext, mockCallback);
            } catch (err) {
                expect(err).toEqual(new Error('Unknown TFL API'));
            }
        });

        it ('should return exception when the fetched secret is not as expected', async () => {
            getSecretValueMock.mockReturnValueOnce({
                promise: () => ({
                    SecretString: "{\"empty\" : true}"
                })
            });
            try {
                const tglApiEnv = 'TFL_API_Portal';
                process.env.TFL_API_SECRET_ID = tglApiEnv;
                await handler(theEvent, theContext, mockCallback);
            } catch (err) {
                expect(err).toEqual(new Error('Unexpected TFL API credentials'));
            }
        });

        it('should call the handler with success', async () => {
            const tflApiEnv = 'TFL_API_Portal';
            process.env.TFL_API_SECRET_ID = tflApiEnv;

            getSecretValueMock.mockReturnValue({
                promise: () => ({
                    SecretString: "{\"tfl_api_app_id\" : 123, \"tfl_api_app_key\" : 456}"
                })
            });

            theEvent.destination =  "Croydon";

            const returnVal = await handler(theEvent, theContext, mockCallback);
            const theBody = JSON.parse(returnVal.body);

            expect(getSecretValueMock).toHaveBeenCalledWith({SecretId: tflApiEnv});
            expect(theBody.message).toEqual(`WOZiTech nextBusto ${theEvent.destination}`)
            expect(returnVal.arrivals.length).toEqual(3);
            expect(returnVal.arrivals[1]).toEqual('2018-11-27T09:44:02Z');
            expect(returnVal).toMatchSnapshot();
        });
            
    });

    describe('Calling the next bus model', () => {
        afterEach(() => {
            jest.clearAllMocks();
        });

        // Need to mock "nextBusTo"
        // test conditions:
        //  1. unknown destination - returns success, but the response.err is set
        //  2. exception (simulates an Axios.get error - could be gateway et al)

    });
});