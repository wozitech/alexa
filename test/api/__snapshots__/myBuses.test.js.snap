// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The myBuses handler Calling the handler should call the handler with success 1`] = `
Object {
  "arrivals": Array [
    "2018-11-27T11:13:33Z",
    "2018-11-27T09:44:02Z",
    "2018-11-27T11:19:05Z",
  ],
  "body": "{\\"message\\":\\"WOZiTech nextBusto Croydon\\",\\"event\\":{\\"destination\\":\\"Croydon\\"}}",
  "statusCode": 200,
}
`;
